<img src="http://140.116.53.170/classes/numericalOptimization/simplexMethods/figure001.png" width="600" alt="Given Information">

In [1]:
import numpy as np
def E1(A,i,k):
    m=len(A)
    n=len(A[0,:])
    I=np.identity(m)
    I[i,i]=k
    return I.dot(A)
def E2(A,i,j,k):
    m=len(A)
    n=len(A[0,:])
    I=np.identity(m)
    I[j,i]=k
    return I.dot(A)


In [2]:
np.set_printoptions(precision=3,suppress=True)
A=np.zeros((4,8))
A[0,0]=1
A[0,1]=-3
A[0,2]=-5
A[0,3]=0
A[0,4]=0
A[0,5]=0
A[0,6]=0

A[1,0]=0
A[1,1]=1
A[1,2]=0
A[1,3]=1
A[1,4]=0
A[1,5]=0
A[1,6]=4

A[2,0]=0
A[2,1]=0
A[2,2]=2
A[2,3]=0
A[2,4]=1
A[2,5]=0
A[2,6]=12

A[3,0]=0
A[3,1]=3
A[3,2]=2
A[3,3]=0
A[3,4]=0
A[3,5]=1
A[3,6]=18

print(A)
m=len(A)
n=len(A[0])
print(m,n)

rq=-1
while rq<0: 
    #######################################
    rq=float("inf")
    j=-1
    for l in range(1,n-2):
        if A[0,l]<rq:
            rq=A[0,l]
            j=l
    print(rq,j)
    #######################################    
    if rq>=0:
        break
    #######################################
    MRT=float("inf")
    i=-1
    for l in range(1,m):
        if A[l,j]>0:
            A[l,n-1]=A[l,n-2]/A[l,j]
            if A[l,n-2]/A[l,j]<MRT:
                MRT=A[l,n-2]/A[l,j]
                i=l
    print(MRT,i)
    print(A)
    #######################################
    A=E1(A,i,1/A[i,j])
    A=E2(A,i,0,-A[0,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    #for l in range(0,i):    
    #    A=E2(A,i,0,-A[l,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    for l in range(i+1,m):
        A=E2(A,i,l,-A[l,j])
    #######################################
    for i in range(m):
        A[i,n-1]=0
    
    #print(A)
    
print(A)    

[[ 1. -3. -5.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  0.]
 [ 0.  3.  2.  0.  0.  1. 18.  0.]]
4 8
-5.0 2
6.0 2
[[ 1. -3. -5.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  6.]
 [ 0.  3.  2.  0.  0.  1. 18.  9.]]
-3.0 1
2.0 3
[[ 1.  -3.   0.   0.   2.5  0.  30.   0. ]
 [ 0.   1.   0.   1.   0.   0.   4.   4. ]
 [ 0.   0.   1.   0.   0.5  0.   6.   0. ]
 [ 0.   3.   0.   0.  -1.   1.   6.   2. ]]
0.0 1
[[ 1.     0.     0.     0.     1.5    1.    36.     0.   ]
 [ 0.     0.     0.     1.     0.333 -0.333  2.     0.   ]
 [ 0.     0.     1.     0.     0.5    0.     6.     0.   ]
 [ 0.     1.     0.     0.    -0.333  0.333  2.     0.   ]]


HTML Tables

HTML tables allow web developers to arrange data into rows and columns.

<a href="https://www.w3schools.com/html/html_tables.asp">HTML Tables</a>


In [3]:
#Reading the following cell in order to obtain some infromation related to the table in html

<h2>Simplex Tableau</h2>
<p>Setp 0:</p>
<table class="table table-bordered table-hover "><tbody><tr>
<th></th><th style="text-align:center">Z </th><th style="text-align:center">$x_1$ </th><th style="text-align:center">$x_2$ </th><th style="text-align:center">$x_3$ </th><th style="text-align:center">$x_4$ </th><th style="text-align:center">$x_5$ </th><th style="text-align:center">RHS </th><th style="text-align:center"> </th></tr> 
<tr><td style="text-align:center">Z </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> -3.00  </td><td style="text-align:center"> -5.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> </td></tr> 
<tr><td style="text-align:center">$x_3$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 4.00  </td><td style="text-align:center">-- </td></tr> 
<tr><td style="text-align:center">$x_4$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 2.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 12.00  </td><td style="text-align:center"> 6.00  </td></tr> 
<tr><td style="text-align:center">$x_5$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 3.00  </td><td style="text-align:center"> 2.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 18.00  </td><td style="text-align:center"> 9.00  </td></tr> 
</tbody></table>
<ul>
<li>
The value of reduced cost of the variable $x_2$ is -5.
</li>
<li>
The value of minimum ratio test is the minimum of {--, 6, 9} = 6.
</li>
<li>
Hence the variable $x_2$ should leave the basis.
</li>
 
 
 
</ul>
<p>Setp 1:</p>
<table class="table table-bordered table-hover "><tbody><tr>
<th></th><th style="text-align:center">Z </th><th style="text-align:center">$x_1$ </th><th style="text-align:center">$x_2$ </th><th style="text-align:center">$x_3$ </th><th style="text-align:center">$x_4$ </th><th style="text-align:center">$x_5$ </th><th style="text-align:center">RHS </th><th style="text-align:center"> </th></tr> 
<tr><td style="text-align:center">Z </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> -3.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 2.50  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 30.00  </td><td style="text-align:center"> </td></tr> 
<tr><td style="text-align:center">$x_3$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 4.00  </td><td style="text-align:center"> 4.00  </td></tr> 
<tr><td style="text-align:center">$x_2$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.50  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 6.00  </td><td style="text-align:center">-- </td></tr> 
<tr><td style="text-align:center">$x_5$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 3.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> -1.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 6.00  </td><td style="text-align:center"> 2.00  </td></tr> 
</tbody></table>
<ul>
<li>
The value of reduced cost of the variable $x_1$ is -3.
</li>
<li>
The value of minimum ratio test is the minimum of {4, --, 2} = 2.
</li>
<li>
Hence the variable $x_1$ should leave the basis.
</li>
 
 
 
</ul>
<p>Setp 2:</p>
<table class="table table-bordered table-hover "><tbody><tr>
<th></th><th style="text-align:center">Z </th><th style="text-align:center">$x_1$ </th><th style="text-align:center">$x_2$ </th><th style="text-align:center">$x_3$ </th><th style="text-align:center">$x_4$ </th><th style="text-align:center">$x_5$ </th><th style="text-align:center">RHS </th><th style="text-align:center"> </th></tr> 
<tr><td style="text-align:center">Z </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.50  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 36.00  </td></tr> 
<tr><td style="text-align:center">$x_3$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.33  </td><td style="text-align:center"> -0.33  </td><td style="text-align:center"> 2.00  </td></tr> 
<tr><td style="text-align:center">$x_2$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.50  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 6.00  </td></tr> 
<tr><td style="text-align:center">$x_1$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> -0.33  </td><td style="text-align:center"> 0.33  </td><td style="text-align:center"> 2.00  </td></tr> 
</tbody></table>
<ul>
<li>
The reduced cost vector is nonnegative.</li>
<li>
The optimal solution is ($x_1$, $x_2$, $x_3$, $x_4$, $x_5$) = (2.00, 6.00, 2.00, 0, 0)
</li>
</ul>

<a href="http://140.116.53.170/classes/numericalOptimization/simplexMethods/simplexTableau.html">Simplex Tableau Example</a>

<a href="simplexTabuleau.txt">Simplex Tableau Example</a>

In [4]:
f = open("SimplexTabuleau.txt", "w")
f.write("Now the file has more content!")
f.close()

In [5]:
f = open("SimplexTabuleau.html", "w")
f.write("Now the file has more content!")


f.write("<table>")
f.write("  <tr>")
f.write("    <th>")
f.write(str(A[1,1]))
f.write("</th>")
f.write("    <th>Contact</th>")
f.write("    <th>Country</th>")
f.write("  </tr>")
f.write("</table>")
f.close()
print(A[1,1])

0.0


Open file and return a corresponding file object. If the file cannot be opened, an OSError is raised. See Reading and Writing Files for more examples of how to use this function.

<a href="https://docs.python.org/3/library/functions.html#open">open()</a>

<a href="https://docs.python.org/3/tutorial/inputoutput.html#tut-files">Reading and Writing Files</a>



In [6]:
class LPOption:
    def __init__(self):
        self.Maximize = 0
        self.Minimize = 0
        self.ConstraintType=[]
        self.VariableUnresticted=[]

In [7]:
def solveLP(init_A,init_B,init_C,optimalType,constrainType,initialSolution):
    
    M = 20000
    
    if initialSolution != True:  #還沒有初始解
        #lt eq gt is constrain type
        lt,eq,gt = 0,0,0
        for i in constrainType:
            if i == 0:
                lt = lt + 1
            elif i == 1:
                eq = eq + 1
            elif i == 2:
                gt = gt + 1
        
        m = 1 + len(init_A)
        n = 1 + len(init_A[0]) + 1 + 1 + lt + eq + gt * 2
        np.set_printoptions(precision=3,suppress=True)
        #print(lt,eq,gt)
        #print(m,n)
        A = np.zeros((m,n))
        
        
    
        
        if optimalType == 1:   #max problem
            A[0,0] = 1
            for i in range(len(init_C)):
                A[0,1+i] = -init_C[i][0]
        else:  #min problem
            A[0,0] = -1
            for i in range(len(init_C)):
                A[0,1+i] = init_C[i][0]
            
        for i in range(len(init_A)):
            for j in range(len(init_A[0])):
                A[i+1,j+1] = init_A[i][j]
        for i in range(len(init_B)):
            A[i+1,-2] = init_B[i][0]
            
            
        #set basic variable BV
        t = 1 + len(init_A[0]) 
        for i in range(len(init_A)):
            if constrainType[i] == 0:
                A[i+1,t] = 1
                
                t = t + 1
            elif constrainType[i] == 1:
                A[i+1,t] = 1
                A[0,t] = M
                A = E2(A,i+1,0,-M)
                t = t + 1
            elif constrainType[i] == 2:
                A[i+1,t] = -1
                A[i+1,t+1] = 1
                A[0,t+1] = M
                A = E2(A,i+1,0,-M)
                t = t + 2
        print("Initial Tabuleau :")
        
        print(A)
    else:  #已經有初始解
        lt,eq,gt = 0,0,0
        for i in constrainType:
            if i == 0:
                lt = lt + 1
            elif i == 1:
                eq = eq + 1
            elif i == 2:
                gt = gt + 1
        
        m = 1 + len(init_A)
        n = 1 + len(init_A[0]) + 1 + 1 
        #print(lt,eq,gt)
        #print(m,n)
        A = np.zeros((m,n))
        if optimalType == 1:   #max problem
            A[0,0] = 1
            for i in range(len(init_C)):
                A[0,1+i] = -init_C[i][0]
        else:  #min problem
            A[0,0] = -1
            for i in range(len(init_C)):
                A[0,1+i] = init_C[i][0]
            
        for i in range(len(init_A)):
            for j in range(len(init_A[0])):
                A[i+1,j+1] = init_A[i][j]
        for i in range(len(init_B)):
            A[i+1,-2] = init_B[i][0]
            
            
        print("Initial Tabuleau :")
        print(A)

    num = 1
    
    rq=-1
    while rq<0: 
        print("########Iteration " + str(num) +"########")
        num += 1
        #######################################
        rq=float("inf")
        j=-1
        for l in range(1,n-2):
            if A[0,l]<rq:
                rq=A[0,l]
                j=l
        #print(rq,j)
        if rq == 0:
            print("The end tabuleau!!")
            break
        else:    
            print("Choose " + str(rq) + ", at column " + str(j))
        #######################################    
        if rq>=0:
            break
        #######################################
        MRT=float("inf")
        i=-1
        for l in range(1,m):
            if A[l,j]>0:
                A[l,n-1]=A[l,n-2]/A[l,j]
                if A[l,n-2]/A[l,j]<MRT:
                    MRT=A[l,n-2]/A[l,j]
                    i=l
        #print(MRT,i)
        print("min = " + str(MRT) + ", at row " + str(i))
        print(A)
        #######################################
        A=E1(A,i,1/A[i,j])
        A=E2(A,i,0,-A[0,j])
        for l in range(1,i):
            A=E2(A,i,l,-A[l,j])
        #for l in range(0,i):    
        #    A=E2(A,i,0,-A[l,j])
        for l in range(1,i):
            A=E2(A,i,l,-A[l,j])
        for l in range(i+1,m):
            A=E2(A,i,l,-A[l,j])
        #######################################
        for i in range(m):
            A[i,n-1]=0

        #print(A)

    print(A)  
    if optimalType == 1:
        print("Optimal value : " + str(A[0,-2]))
    else:
        print("Optimal value : " + str(-A[0,-2]))
    


# Exercise:

# Problem 1

Economical Diet Proposal:
<ol>
<li> There are available 4 different foods at the market.</li>
<li> There are 3 basic nutritional ingredients.</li>
<li> In order to achieve a balanced diet, each individual must receive at least 800g, 90g, 50g of the first, second, third nutrient per day.</li>
<li> The first, second, third, forth food are sold at prices 200NTD, 300NTD, 400NTD, 250NTD per unit.</li>
<li> The relation between food and nutrient is describes as follows.</li>
<li> Determine the most economical diet that satisfies the minimum nutritional requirements.</li>
</ol>

<table class="table table-bordered table-hover "><tbody>
    <tr>
        <td style="text-align:center"> </td>
        <td style="text-align:center">1 </td>
        <td style="text-align:center">2 </td>
        <td style="text-align:center">3 </td>
        <td style="text-align:center">4 </td>
        <td style="text-align:center">minimum requirements </td>    
    </tr>
    <tr>
        <td style="text-align:center">1</td>
        <td style="text-align:center">0.95 </td>
        <td style="text-align:center">0.85 </td>
        <td style="text-align:center">0.75 </td>
        <td style="text-align:center">0.99 </td>
        <td style="text-align:center">800 </td>    
    </tr>
    <tr>
        <td style="text-align:center">2</td>
        <td style="text-align:center">0.025 </td>
        <td style="text-align:center">0.1 </td>
        <td style="text-align:center">0.2 </td>
        <td style="text-align:center">0.01 </td>
        <td style="text-align:center">90</td>    
    </tr>
    <tr>
        <td style="text-align:center">3</td>
        <td style="text-align:center">0.025 </td>
        <td style="text-align:center">0.15 </td>
        <td style="text-align:center">0.05 </td>
        <td style="text-align:center">0 </td>
        <td style="text-align:center">50 </td>    
    </tr>
    <tr>
        <td style="text-align:center">Price of food </td>
        <td style="text-align:center">200</td>
        <td style="text-align:center">300</td>
        <td style="text-align:center">400</td>
        <td style="text-align:center">250</td>
        <td style="text-align:center"> </td>    
    </tr>
</tbody></table>

$$
\begin{array}{rl}
\min & 200x_1+300x_2+400x_3+250x_4\\
s.t. & 0.95x_1+0.85x_2+0.75x_3+0.99x_4\geq 800\\
~& 0.025x_1+0.1x_2+0.2x_3+0.01x_4\geq 90\\
~& 0.025x_1+0.15x_2+0.05x_3\geq 50\\
~& x_1\geq 0,~x_2\geq 0,~x_3\geq 0,~x_4\geq 0\\
\end{array}
$$

In [8]:
#Problem 1
np.set_printoptions(precision=3,suppress=True)
m,n=3,4
A=np.zeros((m,n))
b=np.zeros((m,1))
c=np.zeros((n,1))
A[0,0]=0.95
A[0,1]=0.85
A[0,2]=0.75
A[0,3]=0.99
A[1,0]=0.025
A[1,1]=0.1
A[1,2]=0.2
A[1,3]=0.01
A[2,0]=0.025
A[2,1]=0.15
A[2,2]=0.05
A[2,3]=0
print(A)

b[0,0]=800
b[1,0]=90
b[2,0]=50
print(b)

c[0,0]=200
c[1,0]=300
c[2,0]=400
c[3,0]=250
print(c)

LP1 = LPOption()
LP1.Maximize=0
LP1.Minimize=1
LP1.ConstraintType=[2,2,2]##since all constraints are ineqality constraints and the ineqality signs are ``>=''.
LP1.VariableUnresticted=[0,0,0,0]#since all variables are nonnegative.
solveLP(A,b,c,LP1.Maximize,LP1.ConstraintType,False)

[[0.95  0.85  0.75  0.99 ]
 [0.025 0.1   0.2   0.01 ]
 [0.025 0.15  0.05  0.   ]]
[[800.]
 [ 90.]
 [ 50.]]
[[200.]
 [300.]
 [400.]
 [250.]]
Initial Tabuleau :
[[       -1.       -19800.       -21700.       -19600.       -19750.
      20000.            0.        20000.            0.        20000.
          0.    -18800000.            0.   ]
 [        0.            0.95          0.85          0.75          0.99
         -1.            1.            0.            0.            0.
          0.          800.            0.   ]
 [        0.            0.025         0.1           0.2           0.01
          0.            0.           -1.            1.            0.
          0.           90.            0.   ]
 [        0.            0.025         0.15          0.05          0.
          0.            0.            0.            0.           -1.
          1.           50.            0.   ]]
########Iteration 1########
Choose -21700.0, at column 2
min = 333.33333333333337, at row 3
[[       -1.

# Problem 2
$$
\begin{array}{rl}
\max & 3x_1+5x_2\\
s.t.   & x_1+x_3=4\\
       & 2x_2+x_4=12\\
       & 3x_1+2x_2+x_5=18\\
       & x_1,x_2,x_3,x_4,x_5\geq0\\
\end{array}$$

In [9]:
#Problem 2
np.set_printoptions(precision=3,suppress=True)
m=3
n=5

A=np.zeros((m,n))
A[0,0]=1
A[0,2]=1

A[1,1]=2
A[1,3]=1

A[2,0]=3
A[2,1]=2
A[2,4]=1
print(A)

b=np.zeros((m,1))
b[0,0]=4
b[1,0]=12
b[2,0]=18
print(b)

c=np.zeros((n,1))
c[0,0]=3
c[1,0]=5
print(c)


LP2 = LPOption()
LP2.Maximize=1
LP2.Minimize=0
LP2.ConstraintType=[1,1,1]#since all constraints are eqality constraints.
LP2.VariableUnresticted=[0,0,0,0,0]#since all variables are nonnegative.
solveLP(A,b,c,LP2.Maximize,LP2.ConstraintType,True)

[[1. 0. 1. 0. 0.]
 [0. 2. 0. 1. 0.]
 [3. 2. 0. 0. 1.]]
[[ 4.]
 [12.]
 [18.]]
[[3.]
 [5.]
 [0.]
 [0.]
 [0.]]
Initial Tabuleau :
[[ 1. -3. -5. -0. -0. -0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  0.]
 [ 0.  3.  2.  0.  0.  1. 18.  0.]]
########Iteration 1########
Choose -5.0, at column 2
min = 6.0, at row 2
[[ 1. -3. -5. -0. -0. -0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  6.]
 [ 0.  3.  2.  0.  0.  1. 18.  9.]]
########Iteration 2########
Choose -3.0, at column 1
min = 2.0, at row 3
[[ 1.  -3.   0.   0.   2.5  0.  30.   0. ]
 [ 0.   1.   0.   1.   0.   0.   4.   4. ]
 [ 0.   0.   1.   0.   0.5  0.   6.   0. ]
 [ 0.   3.   0.   0.  -1.   1.   6.   2. ]]
########Iteration 3########
The end tabuleau!!
[[ 1.     0.     0.     0.     1.5    1.    36.     0.   ]
 [ 0.     0.     0.     1.     0.333 -0.333  2.     0.   ]
 [ 0.     0.     1.     0.     0.5    0.     6.     0.   ]
 [ 0.     1.     0.     0.    -0.333  0.333  2. 

# Problem 3
$$\begin{array}{rl}
\max & 3x_1+5x_2+6x_3\\
s.t.   & 2x_1+x_2+x_3\leq 4\\
       & x_1+2x_2+x_3\leq 4\\
       & x_1+x_2+2x_3\leq 4\\
       & x_1+x_2+2x_3\leq 3\\
       & x_1,x_2,x_3\geq0\\
\end{array}$$

In [10]:
#Problem 3
np.set_printoptions(precision=3,suppress=True)
m=4
n=3

A=np.zeros((m,n))
A[0,0]=2
A[0,1]=1
A[0,2]=1

A[1,0]=1
A[1,1]=2
A[1,2]=1

A[2,0]=1
A[2,1]=1
A[2,2]=2

A[3,0]=1
A[3,1]=1
A[3,2]=2
print(A)

b=np.zeros((m,1))
b[0,0]=4
b[1,0]=4
b[2,0]=4
b[3,0]=3
print(b)

c=np.zeros((n,1))
c[0,0]=3
c[1,0]=5
c[2,0]=6
print(c)


LP3 = LPOption()
LP3.Maximize=1
LP3.Minimize=0
LP3.ConstraintType=[0,0,0,0]#since all constraints are ineqality constraints and the ineqality signs are ``<=''.
LP3.VariableUnresticted=[0,0,0]#since all variables are nonnegative.
solveLP(A,b,c,LP3.Maximize,LP3.ConstraintType,False)

[[2. 1. 1.]
 [1. 2. 1.]
 [1. 1. 2.]
 [1. 1. 2.]]
[[4.]
 [4.]
 [4.]
 [3.]]
[[3.]
 [5.]
 [6.]]
Initial Tabuleau :
[[ 1. -3. -5. -6.  0.  0.  0.  0.  0.  0.]
 [ 0.  2.  1.  1.  1.  0.  0.  0.  4.  0.]
 [ 0.  1.  2.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  1.  1.  2.  0.  0.  1.  0.  4.  0.]
 [ 0.  1.  1.  2.  0.  0.  0.  1.  3.  0.]]
########Iteration 1########
Choose -6.0, at column 3
min = 1.5, at row 4
[[ 1.  -3.  -5.  -6.   0.   0.   0.   0.   0.   0. ]
 [ 0.   2.   1.   1.   1.   0.   0.   0.   4.   4. ]
 [ 0.   1.   2.   1.   0.   1.   0.   0.   4.   4. ]
 [ 0.   1.   1.   2.   0.   0.   1.   0.   4.   2. ]
 [ 0.   1.   1.   2.   0.   0.   0.   1.   3.   1.5]]
########Iteration 2########
Choose -2.0, at column 2
min = 1.6666666666666667, at row 2
[[ 1.     0.    -2.     0.     0.     0.     0.     3.     9.     0.   ]
 [ 0.     1.5    0.5    0.     1.     0.     0.    -0.5    2.5    5.   ]
 [ 0.     0.5    1.5    0.     0.     1.     0.    -0.5    2.5    1.667]
 [ 0.     0.     0.     0.

# Problem 4
$$
\begin{array}{rl}
\min & 0.4x_1+0.5x_2\\
s.t.   & 0.3x_1+0.1x_2\leq 2.7\\
       & 0.5x_1+0.5x_2= 6\\
       & 0.6x_1+0.4x_2\geq 6\\
       & x_1,x_2\geq0\\
\end{array}$$

In [11]:
#Problem 4
np.set_printoptions(precision=3,suppress=True)
m=3
n=2

A=np.zeros((m,n))
A[0,0]=0.3
A[0,1]=0.1

A[1,0]=0.5
A[1,1]=0.5

A[2,0]=0.6
A[2,1]=0.4

print(A)

b=np.zeros((m,1))
b[0,0]=2.7
b[1,0]=6
b[2,0]=6
print(b)

c=np.zeros((n,1))
c[0,0]=0.4
c[1,0]=0.5
print(c)


LP4 = LPOption()
LP4.Maximize=0
LP4.Minimize=1
LP4.ConstraintType=[0,1,2]
LP4.VariableUnresticted=[0,0]#since all variables are nonnegative.
solveLP(A,b,c,LP4.Maximize,LP4.ConstraintType,False)

[[0.3 0.1]
 [0.5 0.5]
 [0.6 0.4]]
[[2.7]
 [6. ]
 [6. ]]
[[0.4]
 [0.5]]
Initial Tabuleau :
[[     -1.   -21999.6  -17999.5       0.        0.    20000.        0.
  -240000.        0. ]
 [      0.        0.3       0.1       1.        0.        0.        0.
        2.7       0. ]
 [      0.        0.5       0.5       0.        1.        0.        0.
        6.        0. ]
 [      0.        0.6       0.4       0.        0.       -1.        1.
        6.        0. ]]
########Iteration 1########
Choose -21999.6, at column 1
min = 9.000000000000002, at row 1
[[     -1.   -21999.6  -17999.5       0.        0.    20000.        0.
  -240000.        0. ]
 [      0.        0.3       0.1       1.        0.        0.        0.
        2.7       9. ]
 [      0.        0.5       0.5       0.        1.        0.        0.
        6.       12. ]
 [      0.        0.6       0.4       0.        0.       -1.        1.
        6.       10. ]]
########Iteration 2########
Choose -10666.3, at column 2
min = 2.

# Problem 5
$$
\begin{array}{rl}
\min & x_1-7x_2+3x_3\\
s.t.   & 2x_1+x_2-x_3\leq 4\\
       & 4x_1-3x_2\leq 2\\
       & -3x_1+2x_2+x_3\leq 3\\
       & x_1,x_2,x_3\geq0\\
\end{array}
$$

In [12]:
#Problem 4
np.set_printoptions(precision=3,suppress=True)
m=3
n=3

A=np.zeros((m,n))
A[0,0]=2
A[0,1]=1
A[0,2]=-1

A[1,0]=4
A[1,1]=-3

A[2,0]=-3
A[2,1]=2
A[2,2]=1

print(A)

b=np.zeros((m,1))
b[0,0]=4
b[1,0]=2
b[2,0]=3
print(b)

c=np.zeros((n,1))
c[0,0]=1
c[1,0]=-7
c[2,0]=3
print(c)


LP5 = LPOption()
LP5.Maximize=0
LP5.Minimize=1
LP5.ConstraintType=[0,0,0]
LP5.VariableUnresticted=[0,0,0]#since all variables are nonnegative.
solveLP(A,b,c,LP5.Maximize,LP5.ConstraintType,False)

[[ 2.  1. -1.]
 [ 4. -3.  0.]
 [-3.  2.  1.]]
[[4.]
 [2.]
 [3.]]
[[ 1.]
 [-7.]
 [ 3.]]
Initial Tabuleau :
[[-1.  1. -7.  3.  0.  0.  0.  0.  0.]
 [ 0.  2.  1. -1.  1.  0.  0.  4.  0.]
 [ 0.  4. -3.  0.  0.  1.  0.  2.  0.]
 [ 0. -3.  2.  1.  0.  0.  1.  3.  0.]]
########Iteration 1########
Choose -7.0, at column 2
min = 1.5, at row 3
[[-1.   1.  -7.   3.   0.   0.   0.   0.   0. ]
 [ 0.   2.   1.  -1.   1.   0.   0.   4.   4. ]
 [ 0.   4.  -3.   0.   0.   1.   0.   2.   0. ]
 [ 0.  -3.   2.   1.   0.   0.   1.   3.   1.5]]
########Iteration 2########
Choose -9.5, at column 1
min = 0.7142857142857143, at row 1
[[-1.    -9.5    0.     6.5    0.     0.     3.5   10.5    0.   ]
 [ 0.     3.5    0.    -1.5    1.     0.    -0.5    2.5    0.714]
 [ 0.    -0.5    0.     1.5    0.     1.     1.5    6.5    0.   ]
 [ 0.    -1.5    1.     0.5    0.     0.     0.5    1.5    0.   ]]
########Iteration 3########
The end tabuleau!!
[[-1.     0.     0.     2.429  2.714  0.     2.143 17.286  0.   ]
 [ 0.